File: javascript/2015/day01a/day01a.js
#!/usr/bin/env node

const fs = require('fs');
const path = require('path');
const process = require('process');

function usage() {
    const progname = path.basename(process.argv[1]);
    console.log(`usage: ${progname} <input file>`);
    process.exit(1);
}

function processFile(filename) {
    let counter = 0;
    const data = fs.readFileSync(filename, 'utf8');
    for ( let i = 0; i < data.length; i++ ) {
        const ch = data[i];
        if ( ch == '(' ) {
            counter++;
        } else if ( ch == ')' ) {
            counter--;
        }
    }
    return counter;
}

function main() {
    if ( process.argv.length < 3 ) {
        usage();
    }
    const filename = process.argv[2];
    const result = processFile(filename);
    console.log(`result = ${result}`);
}

main()

File: julia/2015/day01a/day01a.jl
#!/usr/bin/env julia

function usage()
    progname = PROGRAM_FILE
    println(stderr, "usage: $progname <input file>")
    exit(1)
end

function process(content::String)::Int32
    floor = 0
    for c in content
        if c == '('
            floor += 1
        elseif c == ')'
            floor -= 1
        end
    end
    return Int32(floor)
end

function main()
    args = ARGS
    if length(args) < 1
        usage()
    end
    filename = args[1]
    try
        content = read(filename, String)
        result = process(content)
        println("result = $result")
    catch e
        println(stderr, "error reading file: $e")
        exit(1)
    end
end

if abspath(PROGRAM_FILE) == @__FILE__
    main()
end
